redirect/src/main/java/com/adyen/checkout/redirect/RedirectUtil.kt

```diff
@@ -0,0 +1,117 @@
+    const val REDIRECT_RESULT_SCHEME = BuildConfig.checkoutRedirectScheme + "://"
+
+    private const val PAYLOAD_PARAMETER = "payload"
+    private const val REDIRECT_RESULT_PARAMETER = "redirectResult"
+    private const val PAYMENT_RESULT_PARAMETER = "PaRes"
+    private const val MD_PARAMETER = "MD"
+    private const val QUERY_STRING_RESULT = "returnUrlQueryString"
+
+    private const val RESOLVER_ACTIVITY_PACKAGE_NAME = "android"
+
+    @SuppressWarnings("TooGenericExceptionCaught")
+    private fun determineResolveResult(context: Context, uri: Uri): ResolveResult {
+        return try {
+            val intent = Intent(Intent.ACTION_VIEW, uri)
+            val browserIntent = Intent(Intent.ACTION_VIEW, Uri.parse("http://"))
+            val packageManager = context.packageManager
+            val resolveInfo = packageManager.resolveActivity(intent, 0)
+            val browserInfo = packageManager.resolveActivity(browserIntent, PackageManager.MATCH_DEFAULT_ONLY)
+            val resolvedPackageName = resolveInfo?.activityInfo?.packageName
+            val browserPackageName = browserInfo?.activityInfo?.packageName
+            when (resolvedPackageName) {
+                RESOLVER_ACTIVITY_PACKAGE_NAME -> ResolveResult(ResolveResult.Type.RESOLVER_ACTIVITY, resolveInfo)
+                browserPackageName -> ResolveResult(ResolveResult.Type.DEFAULT_BROWSER, resolveInfo)
+                null -> ResolveResult(ResolveResult.Type.UNKNOWN, null)
+                else -> ResolveResult(ResolveResult.Type.APPLICATION, resolveInfo)
+            }
+        } catch (e: Exception) {
+            Logger.e(TAG, "determineResolveResult exception", e)
+            ResolveResult(ResolveResult.Type.UNKNOWN, null)
+        }
+    }
+
+    /**
+     * A redirect may return to the application using the ReturnUrl when properly setup in an Intent Filter. Is usually contains result information
+     * as parameters on that returnUrl. This method parses those results and returns a [JSONObject] to be used in the details call.
+     *
+     * @param data The returned Uri
+     * @return The parsed value to be passed on the payments/details call, on the details parameter.
+     */
+    @JvmStatic
+    @Throws(CheckoutException::class)
+    fun parseRedirectResult(data: Uri): JSONObject {
+        Logger.d(TAG, "parseRedirectResult - $data")
+
+        val extractedParams = HashMap<String, String>().apply {
+            data.getQueryParameter(PAYLOAD_PARAMETER)?.let { put(PAYLOAD_PARAMETER, it) }
+            data.getQueryParameter(REDIRECT_RESULT_PARAMETER)?.let { put(REDIRECT_RESULT_PARAMETER, it) }
+            data.getQueryParameter(PAYMENT_RESULT_PARAMETER)?.let { paymentResult ->
+                data.getQueryParameter(MD_PARAMETER)?.let { md ->
+                    put(PAYMENT_RESULT_PARAMETER, paymentResult)
+                    put(MD_PARAMETER, md)
+                }
+            }
+            data.encodedQuery?.let { put(QUERY_STRING_RESULT, it) }
```

Comment:
How is this not adding the `QUERY_STRING_RESULT` when one of the other ones is present? From the tests it seems to work but I don't get how.

Reply:
It does add `QUERY_STRING_RESULT` as long as there are any query params, even if we don't recognize them. However if there are no query params at all, `data.encodedQuery` is null and `QUERY_STRING_RESULT` won't be added.